### PR TITLE
RUM-1520 Single Feature Integration Tests: Trace

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Implement the Single-FIT layer of our new Test Pyramid for the trace module.

### CODE REVIEW

File: features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/handlers/AndroidSpanLogsHandler.kt [hunk 1]
```diff
@@ -65,7 +65,7 @@ internal class AndroidSpanLogsHandler(
         timestampMicroseconds: Long? = null
     ) {
         val logsFeature = sdkCore.getFeature(Feature.LOGS_FEATURE_NAME)
-        if (logsFeature != null) {
+        if (logsFeature != null && fields.isNotEmpty()) {
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/handlers/AndroidSpanLogsHandler.kt:68]: Why this change is needed?

First, with this change we can have a false positive warning in LogCat about missing Logs features, when it is actually registered.

Second, we can have then a possibility to not send anything.

File: reliability/stub-core/src/main/kotlin/com/datadog/android/tests/ktx/JsonObjectExt.kt [hunk 2]
```diff
@@ -6,30 +6,103 @@
...
+ *   "foo": {
+ *     "bar" : [
+ *       {
+ *         "spam": 42
+ *       }
+ *     ]
+ *   }
+ * }
+ */
+@Suppress("UnsafeThirdPartyFunctionCall")
```
Comment (0xnm) [inline reliability/stub-core/src/main/kotlin/com/datadog/android/tests/ktx/JsonObjectExt.kt:47]: should we exclude files under `reliability` from this check?

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/AndroidTracerTest.kt [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+            spanId = span.spanIdAsHexString()
+            Thread.sleep(OP_DURATION_MS)
+            span.finish()
+        }
+
+        // Then
+        val eventsWritten = stubSdkCore.eventsWritten(Feature.TRACING_FEATURE_NAME)
+        assertThat(eventsWritten).hasSize(1)
+        val event0 = JsonParser.parseString(eventsWritten[0].eventData) as JsonObject
+        println(event0)
```
Comment (0xnm) [inline reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/AndroidTracerTest.kt:90]: some debugging leftover here and in the tests below

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/SpanExt.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * Returns the span's spanId in hex format.
+ * The [SpanContext.toSpanId] method returns a string in decimal format,
+ * which doesn't match what we send in our events
+ */
+fun Span.spanIdAsHexString(): String {
+    return context().toSpanId().toLong().toHexString()
+}
+
+/**
+ * Returns the span's traceId in hex format.
```
Comment (0xnm) [inline reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/SpanExt.kt:32]: wrong description here and below

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/TraceConfigurationTest.kt [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+            spanId = span.spanIdAsHexString()
+            Thread.sleep(OP_DURATION_MS)
+            span.finish()
+        }
+
+        // Then
+        val eventsWritten = stubSdkCore.eventsWritten(Feature.TRACING_FEATURE_NAME)
+        assertThat(eventsWritten).hasSize(1)
+        val event0 = JsonParser.parseString(eventsWritten[0].eventData) as JsonObject
+        println(event0)
```
Comment (0xnm) [inline reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/TraceConfigurationTest.kt:148]: debugging leftover here and below

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+            Thread.sleep(OP_DURATION_MS)
+            span.setError(fakeThrowable)
+            span.finish()
+        }
+
+        // Then
+        val eventsWritten = stubSdkCore.eventsWritten(Feature.TRACING_FEATURE_NAME)
+        assertThat(eventsWritten).hasSize(1)
+        val event0 = JsonParser.parseString(eventsWritten[0].eventData) as JsonObject
+        println(event0)
```
Comment (0xnm) [inline reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt:92]: debugging leftover

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+            .isEqualTo(stubSdkCore.getDatadogContext().sdkVersion)
+        assertThat(event0.getInt("spans[0].error")).isEqualTo(1)
+        assertThat(event0.getString("spans[0].name")).isEqualTo(fakeOperation)
+        assertThat(event0.getString("spans[0].resource")).isEqualTo(fakeOperation)
+        assertThat(event0.getLong("spans[0].duration")).isBetween(OP_DURATION_NS, fullDuration)
+        assertThat(event0.getString("spans[0].meta.error.type")).isEqualTo(fakeThrowable.javaClass.canonicalName)
+        assertThat(event0.getString("spans[0].meta.error.msg")).isEqualTo(fakeThrowable.message)
+        assertThat(event0.getString("spans[0].meta.error.stack")).isEqualTo(fakeThrowable.stackTraceToString())
+        val eventsReceived = stubSdkCore.eventsReceived(Feature.LOGS_FEATURE_NAME)
+        assertThat(eventsReceived).hasSize(0)
```
Comment (0xnm) [inline reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt:109]: minor: `isEmpty` will give more descriptive error

```suggestion
        assertThat(eventsReceived).isEmpty
```

**Review by 0xnm — COMMENTED** [review]
nice work! I left some questions/suggestions.

**Review by mariusc83 — APPROVED** [review]
Nice job !!

File: features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/handlers/AndroidSpanLogsHandler.kt (hunk 1 above)
Comment (xgouchet) [reply features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/handlers/AndroidSpanLogsHandler.kt:68]: Essentially this is to avoid sending a useless log: if the `fields` map is empty, then we will send a log with a default message (that just says "span log") with no custom attributes, providing no information whatsoever.
But indeed we might get a false positive logcat message, I'll fix that.

File: reliability/stub-core/src/main/kotlin/com/datadog/android/tests/ktx/JsonObjectExt.kt (hunk 2 above)
Comment (xgouchet) [reply reliability/stub-core/src/main/kotlin/com/datadog/android/tests/ktx/JsonObjectExt.kt:47]: Yes that makes more sense indeed.

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt (hunk 7 above)
Comment (xgouchet) [reply reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt:109]: 👍

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt (hunk 6 above)
Comment (xgouchet) [reply reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/UtilitiesTest.kt:92]: 👍

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/TraceConfigurationTest.kt (hunk 5 above)
Comment (xgouchet) [reply reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/TraceConfigurationTest.kt:148]: 👍

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/SpanExt.kt (hunk 4 above)
Comment (xgouchet) [reply reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/SpanExt.kt:32]: 👍

File: reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/AndroidTracerTest.kt (hunk 3 above)
Comment (xgouchet) [reply reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/AndroidTracerTest.kt:90]: 👍

**Review by jonathanmos — APPROVED** [review]
